contracts/Distense.sol

```diff
@@ -1,52 +1,90 @@
 pragma solidity ^0.4.11;
 
-import './lib/SafeMath.sol';
+import './lib/AddressUtils.sol';
+import './lib/StringUtils.sol';
+import './lib/StringArrayUtils.sol';
 
-contract Distense {
-  using SafeMath for uint256;
+import './lib/Approvable.sol';
+import './lib/Ownable.sol';
 
-  uint256 public totalSupply;
-  string public name;
-  string public symbol;
-  uint8 public decimals;  // TODO https://ethereum.stackexchange.com/questions/9256/float-not-allowed-in-solidity-vs-decimal-places-asked-for-token-contract?rq=1
+contract Distense is Ownable, Approvable {
+  using AddressUtils for address;
+  using StringUtils for string;
+  using StringArrayUtils for string[];
 
-  struct Contributor {
-    uint256 balance;
-    string email;
-    bytes8 countryCode;
+  enum ObjectType { tag, commit, tree, blob }
+  struct GitObject {
+    ObjectType objectType;
+    string ipfsHash;
   }
-  
-  mapping(address => Contributor) public contributors;
-  mapping(bytes32 => address) public emailToAddress;
-  
-  event LogContributionReward(address indexed to, uint256 numDID);
 
-  function Distense() {
-    totalSupply = 0;
-    name = "Distense DID";
-    symbol = "DID";
-    decimals = 1;
+  struct Repo {
+    string metaHash;
+    address[] forks;
+    mapping(address => string[]) refNames;
+    mapping(string => mapping(address => string)) refs;
+    mapping(string => GitObject) objects;
```

Comment:
It may be worth it to swap out a lot of these `string` types with `bytes32` since most of them are fixed-width hashes anyways. Just need to replace the library functions I'm using down below.